lib/publiccloud/azure_client.pm

```diff
@@ -26,8 +28,23 @@ has container_registry => sub { get_required_var('PUBLIC_CLOUD_CONTAINER_IMAGES_
 
 sub init {
     my ($self) = @_;
-    $self->vault(publiccloud::vault->new());
-    $self->vault_create_credentials() unless ($self->key_id);
+    if (my $service_url = get_var('PUBLIC_CLOUD_CREDENTIALS_URL')) {
+        my $user = get_required_var('_SECRET_PUBLIC_CLOUD_CREDENTIALS_USER');
+        my $pwd = get_required_var('_SECRET_PUBLIC_CLOUD_CREDENTIALS_PASS');
+        my $url = Mojo::URL->new($service_url)->userinfo("$user:$pwd");
```

Comment:
Why don't you use 'publiccloud::utils:: get_credentials()`?

Reply:
Because `get_credentials `stores the credentials to a file, which is only needed for GCE. Azure and AWS expect the credentials to be set as environment variables, which is exactly what happens here.

Different CSP, different approaches.

Reply:
yes, that was my thinking, but after reading this, I thought of using also UserAgent for get_credentials. So, wait for it. 
Different CSP -> perl magic :D